src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java

```diff
@@ -221,6 +222,52 @@ public void setTotalExperience(int exp) {
         this.experienceTotal = exp;
     }
 
+    /**
+     * {@link EntityPlayer#addExperienceLevel(int)} and
+     * {@link EntityPlayer#removeExperienceLevel(int)} don't update the total
+     * experience. This recalculates it for plugins to properly make use of it.
+     *
+     * @see #onAddExperienceLevels(int, CallbackInfo)
+     */
+    private void recalculateTotalExperience() {
+        if (!this.dontRecalculateExperience) {
+            this.experienceTotal = this.xpAtLevel(this.experienceLevel) + Math.round(this.experience * this.xpBarCap());
```

Comment:
There is an edge case here where you'll potentially go out of sync with the levels, if your experience level has gone down and `Math.round(this.experience * this.xpBarCap()) == this.xpBarCap()` (so you'd actually have the total experience for the _next_ level), you need to either bump this up to the next level, or remove an experience point - I guess the latter is more appropriate here as you're trying to hit a target level.

You also need to recalculate `this.experience` - as you performed a rounding operation, this value will no longer be correct.